File: 039_Combination_Sum/039_1.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<int> tmp;
        dfs(tmp, candidates, target, 0);
        return res;
    }

    void dfs(vector<int> &tmp, vector<int> &candidates, int target, int begin) {
        if (target < 0) return;
        if (target == 0) {
            res.push_back(tmp);
            return;
        }
        for (int i = begin; i < candidates.size(); i++) {
            if (target - candidates[i] < 0) break;
            tmp.push_back(candidates[i]);
            dfs(tmp, candidates, target - candidates[i], i);
            tmp.pop_back();
        }
    }

private:
    vector<vector<int>> res;
};


int main() {
    vector<int> c = {2, 3, 6, 7};
    Solution so;
    for (const vector<int>& v : so.combinationSum(c, 7)) {
        for (int i : v)
            cout << i << " ";
        cout << "\n";
    }
}

File: 039_Combination_Sum/039_1.py
from typing import List, Dict
from queue import Queue
from collections import deque, Counter
import math
from heapq import heappush, heappushpop


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        def dfs(tmp, target, begin):
            if target < 0:
                return
            if target == 0:
                res.append(tmp)
                return
            for i in range(begin, len(candidates)):
                if target < candidates[i]:
                    break
                dfs(tmp + [candidates[i]], target - candidates[i], i)

        candidates.sort()
        tmp = []
        res = []
        dfs(tmp, target, 0)
        return res


if __name__ == "__main__":
    s = Solution()
    candidates = [2,3,6,7]
    target = 7
    print(s.combinationSum(candidates, target))
